### PR TITLE
fix(YAML): improve matching for keys

### LINKED ISSUE #2058: yaml not rendered correctly
Hi, 

I'm having a problem with my yaml files.

Example yaml file:
```yaml
sidebar:
  Basic Tutorial:
  - Getting Started: docs/getting-started.md
  - Customizing The Sidebar: docs/customizing-the-sidebar.md
  - Customizing The Public Api: docs/customizing-the-public-api.md
  - Discussions: docs/discussions.md
  - Themes and CSS: docs/themes-and-css.md
  More Info:
  - Autodoc Settings: docs/autodocSettings-full-list.md
  - Building Version Documentation: docs/building-version-documentation.md
  - Markdown Cheat Sheet: docs/markdown-cheat-sheet.md
  - Changelog: docs/changelog.md
  - Contact Us: docs/contact-us.md
  - Feature Requests: https://github.com/erez-o/doxiz/issues
autodocSettings: {}
```

rendered as:
yaml highlight

Using the latest version (9.15.8), also made a fiddle http://jsfiddle.net/f261xec4/

Thx

### PR DESCRIPTION
**Pull request by cdbax**
Fixes #2058 #1681 

I had a look at the YAML spec as part of fixing this, and what little I understood was horrifying (did you know JSON is valid YAML!?) so I opted for what is hopefully a nice middle-ground that seems to me to be an improvement on the existing detection.

**BEFORE**
image

**AFTER**
image

### CODE REVIEW

File: src/languages/yaml.js [hunk 1]
```diff
@@ -8,14 +8,15 @@ Category: common, config
...
+  // ...and ending with a colon followed immediately by a space, tab or newline.
+  // The YAML spec allows for much more than this, but this covers most use-cases.
   var KEY = {
     className: 'attr',
     variants: [
-      { begin: keyPrefix + keyName + ":"},
-      { begin: keyPrefix + '"' + keyName + '"' + ":"},
-      { begin: keyPrefix + "'" + keyName + "'" + ":"}
+      { begin: '\\w[\\w :\\/\\.]*:(?=[ \t]|$)' },
+      { begin: '"\\w[\\w :\\/\\.]*":(?=[ \t]|$)' } //quoted keys
```
Comment (yyyc514) [inline src/languages/yaml.js]: What happened to single quoted keys?

File: test/markup/tap/yaml-block.expect.txt [hunk 2]
```diff
@@ -8,10 +8,10 @@
...
 <span class="hljs-symbol">ok</span>
   <span class="yaml"><span class="hljs-meta">---</span>
-<span class="hljs-attr">  message:</span> <span class="hljs-string">"Board layout"</span>
-<span class="hljs-attr">  severity:</span> <span class="hljs-string">comment</span>
-<span class="hljs-attr">  dump:</span>
-<span class="hljs-attr">     board:</span>
+  <span class="hljs-attr">message:</span> <span class="hljs-string">"Board layout"</span>
+  <span class="hljs-attr">severity:</span> <span class="hljs-string">comment</span>
+  <span class="hljs-attr">dump:</span>
+     <span class="hljs-attr">board:</span>
```
Comment (yyyc514) [inline test/markup/tap/yaml-block.expect.txt]: Ha, this is an improvement me thinks :)

File: test/markup/yaml/keys.txt [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+some_key:
+  another_key: value
+
+some key:
+  another key: value
+
+"some key":
+  "another key": value
+
```
Comment (yyyc514) [inline test/markup/yaml/keys.txt]: Would be nice to see "some-key" and "some.key" since those were points of issue from before.

Comment (cdbax) [reply test/markup/yaml/keys.txt]: 👍 Added

File: src/languages/yaml.js (hunk 1 above)
Comment (cdbax) [reply src/languages/yaml.js]: Whoops! Thanks for the save.

**Comment by yyyc514** [comment]
If not it's ok, but it'd be nice to solve that also if you're game.

File: src/languages/yaml.js (hunk 1 above)
Comment (yyyc514) [reply src/languages/yaml.js]: That's what I'm here for.

**Comment by cdbax** [comment]
Yeah I planned to go looking for some of the other YAML issues next. Happy to try and do it in the one PR if that's preferred.

**Comment by yyyc514** [comment]
No, lets split it up if you don't mind.

**Comment by cdbax** [comment]
@yyyc514 can you explain the `|$` hack to me? I could not get lookaheads to work properly until I noticed that other lookaheads in the code seemed to be doing this...but I have no idea why it works.

**Comment by yyyc514** [comment]
It's how the parser works... it's two stages and the second stage doesn't have access to the FULL content, only the match itself... and matches don't include look ahead... so the second stage will fail and the whole thing falls to pieces, unless you say look-ahead or "end of string" because in the 2nd stage all you'v got is the end of string.

And in many cases this hack is "good enough".

I'm currently working on fixing the parser so this isn't necessary and true look-ahead "just works" without caveats, but I've gotten a bit distracted as of late with the larger Highlight.js project. :-)

I plan to return to that work soon though.

**Comment by yyyc514** [comment]
Can you fix the merge conflict?

@egor-rogov  @marcoscaceres You think there is any use in a two reviewer policy for PRs or if one of us reviews (who didn't also write) and then thinks it's good, then merge it?

**Comment by marcoscaceres** [comment]
One reviewer is fine as long as you have a high degree of confidence (and there are no crypto-miners 🤣).

**Comment by yyyc514** [comment]
@marcoscaceres Ok I'll follow that rule then and if I think it needs extra eyeballs I'll ping you all.

**Pull request merged**